cli/cmd/add/component-version/cmd.go

```diff
@@ -0,0 +1,381 @@
+	return prov.pluginManager.ResourcePluginRegistry.GetResourcePlugin(ctx, resource.Access)
+}
+
+func (prov *constructorProvider) GetTargetRepository(ctx context.Context, _ *constructorruntime.Component) (constructor.TargetRepository, error) {
+	plugin, err := prov.pluginManager.ComponentVersionRepositoryRegistry.GetPlugin(ctx, prov.targetRepoSpec)
+	if err != nil {
+		return nil, fmt.Errorf("getting plugin for repository %q failed: %w", prov.targetRepoSpec, err)
+	}
+	var creds map[string]string
+	identity, err := plugin.GetComponentVersionRepositoryCredentialConsumerIdentity(ctx, prov.targetRepoSpec)
+	if err == nil {
+		if creds, err = prov.graph.Resolve(ctx, identity); err != nil {
+			return nil, fmt.Errorf("getting credentials for repository %q failed: %w", prov.targetRepoSpec, err)
+		}
+	} else {
+		slog.WarnContext(ctx, "could not get credential consumer identity for component version repository", "repository", prov.targetRepoSpec, "error", err)
+	}
+	repo, err := plugin.GetComponentVersionRepository(ctx, prov.targetRepoSpec, creds)
+	if err != nil {
+		return nil, fmt.Errorf("getting component version repository for %q failed: %w", prov.targetRepoSpec, err)
+	}
+	return repo, err
+}
+
+func registerConstructorProgressTracker(cmd *cobra.Command, options constructor.Options) (opts constructor.Options, stop func(), err error) {
+	format, err := enum.Get(cmd.Flags(), log.FormatFlagName)
+	if err != nil {
+		return opts, nil, fmt.Errorf("failed to get the log format from the command flag: %w", err)
+	}
+
+	switch format {
+	case log.FormatText:
+		pw := progress.NewWriter()
+		pw.SetOutputWriter(cmd.OutOrStdout())
+		pw.SetUpdateFrequency(100 * time.Millisecond)
+		pw.SetAutoStop(false)
+		var trackers ocmsync.Map[string, *progress.Tracker]
+		options.OnStartComponentConstruct = func(_ context.Context, component *constructorruntime.Component) error {
+			key := component.Name + "/" + component.Version
+			tracker := &progress.Tracker{
+				Message: "component " + key,
+				Total:   1,
+				Units: progress.Units{
+					Formatter: func(value int64) string {
+						base := fmt.Sprintf("%d component version", value)
+						if value > 1 {
+							base += "s"
+						}
+						return base
+					},
+				},
+			}
+			trackers.Store(key, tracker)
+			pw.AppendTracker(tracker)
+			return nil
+		}
+		options.OnEndComponentConstruct = func(_ context.Context, descriptor *descriptor.Descriptor, err error) error {
+			if err != nil {
+				return nil
+			}
+			key := descriptor.Component.Name + "/" + descriptor.Component.Version
+			tracker, ok := trackers.Load(key)
+			if !ok {
+				return fmt.Errorf("tracker for component %q not found", key)
+			}
+			tracker.UpdateMessage(tracker.Message + " constructed")
+			tracker.Increment(1)
+			tracker.MarkAsDone()
+			return nil
+		}
+		// TODO Add Resource and Source tracking in more detail
+		go func() {
+			trackers.Range(func(_ string, tracker *progress.Tracker) bool {
+				tracker.Start()
```

Comment:
Okay, so... You're filling up the tracker in `OnStartComponentConstruct`. But you're immediately starting this go routine. How is not immediately returning? Since the tracker is empty at that point yet until wither OnStartComponentConstruct happens, right?

Reply:
Sure, but the Range is never ranging over anything because this is immediately started and you're only filling the tracker in that OnStartCOmponentConstruct which happens later, right? So this doesn't really do anything other than immediately just call Render with no started trackers.

Reply:
now i got you. youre absolutely right, good catch! thanks!